docs/design/staging-environment-proposal.md

```diff
@@ -0,0 +1,26 @@
+# Chronicle Staging Environment
+
+## High Level Diagram
+
+![Chronicle Staging Diagram](assets/chronicle-staging-diagram.png)
+
+## Overview
+
+1. Github Actions will be used to trigger the workflow for the entire deployment process. It will send a deploy queue entry to a `GCP Cloud Tasks` queue, let’s call it the `deploy-queue` with the necessary metadata for the deploy (docker image tag, git branch, etc). Therefore, we will need a docker registry to maintain the necessary images for the staging instance, we will also have to create a system to remove unused images to reduce cost. We will use the `gcloud` cli to push to the docker registry. Essentially, this Github Actions step will occur after every push to the branch, that way a new environment will essentially be created for that branch on every push.
+
+- `Environment Cleaner` is a service that will be running on the cloud to asynchronously detect branch releases by polling the GitHub API. If a current environment for `branch_1` is still running, but ` branch_1` is deleted, this service will handle the cleanup.
+  - If a branch is deleted, a destroy request to a `GCP Cloud Tasks` queue, let’s call it the `destroy-queue` . The `destroy-queue` entry will then be read by a `GCP Cloud Function` that handles the creation of the environment itself.  
+     \* The reasoning behind this multilayer process is that the `GCP Cloud Function` may have easier access to the core credentials via `GCP Secret Manager` required perform all the necessary installations and deployments. If we can easily inject credentials into the `Environment Cleaner` itself, this step is no longer necessary.
+- Create `deployment.yaml` file and template it for each specific branch. This `deployment.yaml` will the the manifest to deploy each chronicle application to a kubernetes cluster.
```

Comment:
One thing I noticed this document missing is how service URLs are set up and how TLS certificates are going to be issued for these domains.

This is a consequential technical challenge that we should be addressing.

A few things I read up on regarding this problem:

- Google certificates do not support wildcard domains
- Google certificates only support up to 100 domains

We could consider using Fastly for development as well, we can just setup a wildcard domain and use its api to add hosts and point hosts based on the name of the domain